src/eval_utils/run.ts

```diff
@@ -0,0 +1,725 @@
+    PromptLogRequest,
+    PromptRequest,
+    RunStatsResponse,
+    ToolLogRequest,
+    ToolRequest,
+} from "../api";
+import { Flows } from "../api/resources/flows/client/Client";
+import { Prompts } from "../api/resources/prompts/client/Client";
+import { evaluationContext } from "./context";
+import { Dataset, Evaluator, EvaluatorCheck, File, FileResponse } from "./types";
+
+// ANSI escape codes for logging colors
+const YELLOW = "\x1b[93m";
+const CYAN = "\x1b[96m";
+const GREEN = "\x1b[92m";
+const RED = "\x1b[91m";
+const RESET = "\x1b[0m";
+
+type LogResponse =
+    | CreateFlowLogResponse
+    | CreatePromptLogResponse
+    | CreateToolLogResponse
+    | CreateEvaluatorLogResponse;
+
+export function overloadLog<T extends Flows | Prompts>(client: T): T {
+    const originalLog = client.log.bind(client);
+
+    // @ts-ignore
+    const _overloadedLog: T["log"] = async (
+        request: FlowLogRequest | PromptLogRequest,
+        options?: Flows.RequestOptions | Prompts.RequestOptions,
+    ) => {
+        let response: LogResponse | undefined;
+        if (evaluationContext.isEvaluatedFile(request)) {
+            const state = evaluationContext.getState();
+            const { runId, sourceDatapointId, uploadCallback } =
+                evaluationContext.getDatapoint({
+                    inputs: request.inputs,
+                    messages: request.messages,
+                });
+
+            if (request.runId === undefined) {
+                request = {
+                    ...request,
+                    runId,
+                };
+            }
+            if (request.sourceDatapointId === undefined) {
+                request = {
+                    ...request,
+                    sourceDatapointId: sourceDatapointId,
+                };
+            }
+            if (client instanceof Flows) {
+                request = {
+                    ...request,
+                    traceStatus: "complete",
+                };
+            }
+
+            if ("flow" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.flow)) {
+                    response = await originalLog(
+                        {
+                            ...request,
+                            // @ts-ignore Log under the version expected by evaluation, not
+                            // one determined by decorators. Otherwise the evaluation will stale
+                            flow: state?.evaluatedVersion,
+                            output: undefined,
+                            error: `The version of the evaluated Flow must match the version of the callable. Expected: ${JSON.stringify(state!.evaluatedVersion)}, got: ${JSON.stringify(request.flow)}`,
+                        },
+                        options,
+                    );
+                }
+            }
+
+            if ("prompt" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.prompt)) {
```

Comment:
why the `!` assertion?

i think that `!`s should be avoided in general. if code changes such that this assumption can't be made, this wouldn't raise an error (and when it does crash, it's not that useful an error - an error message explaining why we expect state to be defined would be more helpful for a future developer coming to this and wondering why this assertion is here)

Reply:
ack, added explicit check